### PR TITLE
list.h: Change inline function to macro as much as possbile

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
since C89 doesn't support inline keyword: https://github.com/apache/incubator-nuttx/issues/6896

## Impact
Minor

## Testing
Pass CI

### CODE REVIEW

**Review by pkarashchenko — APPROVED** [review]

File: include/nuttx/list.h [hunk 1]
```diff
@@ -299,102 +311,6 @@ list_remove_tail(FAR struct list_node *list)
...
-{
-  return list->next == list;
-}
-
-static inline bool list_is_clear(FAR struct list_node *list)
-{
-  return list->next == NULL;
-}
-
 static inline size_t list_length(FAR struct list_node *list)
```
Comment (pkarashchenko) [inline include/nuttx/list.h:314]: So this will be changed in next P?

Comment (xiaoxiang781216) [reply include/nuttx/list.h:314]: It's hard or impossible if we can't use some compiler extension(e.g. https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html).

File: include/nuttx/list.h [hunk 2]
```diff
@@ -37,8 +37,67 @@
...
+      (list)->prev       = (item); \
+    } \
+  while (0)
+
+#define list_delete(item) \
+  do \
+    { \
+      (item)->next->prev = (item)->prev; \
+      (item)->prev->next = (item)->next; \
+      (item)->prev       = (item)->next = 0; \
```
Comment (pkarashchenko) [inline include/nuttx/list.h:98]:
```suggestion
      (item)->prev       = (item)->next = NULL; \
```

**Review by pkarashchenko — COMMENTED** [review]
In general looks good, but the main difference inline from define is that inlining is done based on compiler decision and with define we have inlining unconditionally.

File: include/nuttx/list.h (hunk 2 above)
Comment (xiaoxiang781216) [reply include/nuttx/list.h]: Done.

**Review by pkarashchenko — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
> In general looks good, but the main difference inline from define is that inlining is done based on compiler decision and with define we have inlining unconditionally.

Yes, it's penalty because we decide to stick C89.

**Review by pkarashchenko — APPROVED** [review]

**Comment by yamt** [comment]
@xiaoxiang781216 
can you clarify the license of macros you added in this PR?
having mostly identical comments as linux makes me nervous.

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 can you clarify the license of macros you added in this PR?

The change come from @Junbo-Zheng:
https://github.com/apache/nuttx/pull/6926/commits/1019e894db526f7d150355f43b0a952b3d631d05
I just help him upstream the patch and append a new change to convert inline functions to macros as much as possible.

> having mostly identical comments as linux makes me nervous.

@Junbo-Zheng could you give some comment on this?

**Comment by xiaoxiang781216** [comment]
Anyway, I rewrite these macro here: https://github.com/apache/nuttx/pull/10357. @yamt please take a look.

**Pull request merged**